src/test/java/perkpack/controllers/CardRestControllerTest.java

```diff
@@ -93,11 +105,101 @@ public void editValidCardTest() throws Exception
         String newName = "Visa";
         String newDescription = "A Better Way";
 
-        //for some reason, I cant do newName or newDescription separately
         mockMvc.perform(patch("/cards/" + savedCard.getId() + "?newName=" + newName + "&newDescription=" + newDescription)).
                 andExpect(status().isOk()).
                 andExpect(jsonPath("$.name", is(newName))).
                 andExpect(jsonPath("$.description", is(newDescription)));
 
     }
+
+    @Test
+    public void addPerkToCardTest() throws Exception
+    {
+        Card savedCard = cardRepository.save(validCard);
+        Category games = new Category("Games");
+        categoryRepository.save(games);
+        Perk p = new Perk("Perk1", "Description for perk", games);
+        perkRepository.save(p);
+
+        mockMvc.perform(patch("/cards/"+savedCard.getId() + "/addPerk/" + p.getName())).
+                andExpect(status().isOk()).
+                andExpect(jsonPath("$.perks[0].name", is(p.getName())));
+    }
+
+    @Test
+    public void removePerkFromCardTest() throws Exception
+    {
+        Card savedCard = cardRepository.save(validCard);
+        Category games = new Category("Games1");
+        categoryRepository.save(games);
+        Perk p = new Perk("Perk2", "Description for perk2", games);
+        perkRepository.save(p);
+
+        mockMvc.perform(patch("/cards/"+savedCard.getId() + "/addPerk/" + p.getName()));
+
+        mockMvc.perform(patch("/cards/"+savedCard.getId() + "/removePerk/" + p.getName())).
+                andExpect(status().isOk()).
+                andExpect(jsonPath("$.perks", hasSize(0)));
+    }
+
+    @Test
+    public void addUserToCardTest() throws Exception
+    {
+        Card savedCard = cardRepository.save(validCard);
+        Category games = new Category("Games2");
+        categoryRepository.save(games);
+        User u = new User("Lava", "Tahir", "lava@gmail.com", "password");
+        userRepository.save(u);
+
+        mockMvc.perform(patch("/cards/"+savedCard.getId() + "/addUser/" + u.getId())).
+                andExpect(status().isOk()).
+                andExpect(jsonPath("$.users[0].firstName", is(u.getFirstName())));
+    }
+
+    @Test
+    public void removeUserFromCardTest() throws Exception
+    {
+        Card savedCard = cardRepository.save(validCard);
+        Category games = new Category("Games3");
+        categoryRepository.save(games);
+        User u = new User("Lava", "Tahir", "lava@gmail.com", "password");
+        userRepository.save(u);
+
+        mockMvc.perform(patch("/cards/"+savedCard.getId() + "/addUser/" + u.getId()));
+
+        mockMvc.perform(patch("/cards/"+savedCard.getId() + "/removeUser/" + u.getId())).
+                andExpect(status().isOk()).
+                andExpect(jsonPath("$.users", hasSize(0)));
+    }
+    /*
+    @Test
+    public void addUserToCardByEmailTest() throws Exception
```

Comment:
Why are these tests commented out?

Reply:
tests correspond to commented out functions in the rest controller

Reply:
Relates to comment below, these will be commented out until the functionality is added.